util/src/main/java/io/kubernetes/client/Exec.java

```diff
@@ -170,20 +171,59 @@ public Process exec(
       throws ApiException, IOException {
     String path = makePath(namespace, name, command, container, stdin, tty);
 
-    WebSocketStreamHandler handler = new WebSocketStreamHandler();
-    ExecProcess exec = new ExecProcess(handler);
+    ExecProcess exec = new ExecProcess();
+    WebSocketStreamHandler handler = exec.getHandler();
     WebSockets.stream(path, "GET", apiClient, handler);
 
     return exec;
   }
 
   private static class ExecProcess extends Process {
-    WebSocketStreamHandler streamHandler;
-    private int statusCode;
+    private final WebSocketStreamHandler streamHandler;
+    private volatile int statusCode;
+    private volatile boolean isDestroyed = false;
 
-    public ExecProcess(WebSocketStreamHandler handler) throws IOException {
-      this.streamHandler = handler;
+    public ExecProcess() throws IOException {
       this.statusCode = -1;
+      this.streamHandler =
+          new WebSocketStreamHandler() {
+            @Override
+            public void close() {
+              if (statusCode == -1) {
+                InputStream inputStream = getInputStream(3);
+                int exitCode = 0;
+                try {

```

Comment:
This code seems fragile. I'd love to see it extracted into a separate method and unit tested.